### PR TITLE
Type checker

### PR DESCRIPTION
**Pull request by S3JER**

### CODE REVIEW

File: GOAT-Compiler/TypeChecker.cs [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+        public override void OutAAssignStmt(AAssignStmt node)
+        {
+            Symbol id = _symbolTable.GetSymbol(node.GetId().Text);
+            if (Convert(node.GetExp(), id.type) == id.type)
+            {
+                _typeDictionary.Add(node, id.type);
+            }
+            else
+            {
+                throw new TypeMismatchException(node);
```
Comment (Casper-NS) [inline GOAT-Compiler/TypeChecker.cs:243]: You already throw this exception in convert...

File: GOAT-Compiler/TypeChecker.cs [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+            }
+        }
+
+        public override void OutAVarDecl(AVarDecl node)
+        {
+            Symbol id = _symbolTable.GetSymbol(node.GetId().Text);
+            if(node.GetExp() != null)
+            {
+                if (Convert(node.GetExp(), id.type) == id.type)
```
Comment (Casper-NS) [inline GOAT-Compiler/TypeChecker.cs:333]: Still a double throw i think

File: SymbolTableTest/TypesOK/FunctionTypes.txt [hunk 3]
```diff
@@ -0,0 +1,4 @@
+﻿float f()
```
Comment (Casper-NS) [inline SymbolTableTest/TypesOK/FunctionTypes.txt:1]: Duplicate of returnIntToFloat

File: SymbolTableTest/TypesOK/ReturnIntToFloat.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+﻿float f()
```
Comment (Casper-NS) [inline SymbolTableTest/TypesOK/ReturnIntToFloat.txt:1]: Duplicate of FunctionTypes

**Review by Casper-NS — COMMENTED** [review]
Fix these, then it looks good

File: GOAT-Compiler/TypeChecker.cs [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+            if (type == Types.Void) 
+            {
+                throw new TypeMismatchException(current);
+            }
+            else
+            {
+                _typeDictionary.Add(current, type);
+            }            
+        }
+        private void EqelNotEqel(Node left, Node right, Node current)
```
[the remark this answers is not in the document]
Comment (S3JER) [reply GOAT-Compiler/TypeChecker.cs:40]: Renamed to EqlNotEqlTypeChecker

File: GOAT-Compiler/TypeChecker.cs (hunk 1 above)
Comment (S3JER) [reply GOAT-Compiler/TypeChecker.cs:243]: Removed

File: GOAT-Compiler/TypeChecker.cs (hunk 2 above)
Comment (S3JER) [reply GOAT-Compiler/TypeChecker.cs:333]: Needs this throw

File: SymbolTableTest/TypesOK/FunctionTypes.txt (hunk 3 above)
Comment (S3JER) [reply SymbolTableTest/TypesOK/FunctionTypes.txt]: @Mati-AAU's problem

File: SymbolTableTest/TypesOK/ReturnIntToFloat.txt (hunk 4 above)
Comment (S3JER) [reply SymbolTableTest/TypesOK/ReturnIntToFloat.txt:1]: @Mati-AAU's problem

**Review by Casper-NS — APPROVED** [review]
lgtm

**Pull request merged**